File: backend/app/schemas/bookings.py
from __future__ import annotations

import uuid
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models import BookingParticipantType, BookingPaymentStatus, BookingSource, BookingStatus
from app.models.enums import BookingRuleAppliesTo, StartLane
from app.schemas.availability import AvailabilityPolicyResult


class BookingCreateParticipantInput(BaseModel):
    participant_type: BookingParticipantType
    person_id: uuid.UUID | None = None
    guest_name: str | None = Field(default=None, max_length=255)
    is_primary: bool = False

    @model_validator(mode="after")
    def validate_shape(self) -> BookingCreateParticipantInput:
        if self.participant_type == BookingParticipantType.GUEST:
            if self.person_id is not None:
                raise ValueError("guest participants cannot include person_id in this phase")
            if not self.guest_name or not self.guest_name.strip():
                raise ValueError("guest_name is required for guest participants")
            self.guest_name = self.guest_name.strip()
            return self
        if self.person_id is None:
            raise ValueError("person_id is required for member and staff participants")
        if self.guest_name is not None:
            raise ValueError("guest_name can only be supplied for guest participants")
        return self


class BookingCreateRequest(BaseModel):
    course_id: uuid.UUID
    tee_id: uuid.UUID | None = None
    start_lane: StartLane | None = None
    slot_datetime: datetime
    slot_interval_minutes: int | None = Field(default=None, ge=1, le=240)
    source: BookingSource = BookingSource.ADMIN
    applies_to: BookingRuleAppliesTo | None = None
    reference_datetime: datetime | None = None
    participants: list[BookingCreateParticipantInput] = Field(default_factory=list, max_length=32)

    @field_validator("slot_datetime", "reference_datetime")
    @classmethod
    def validate_timezone_aware_datetime(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return value
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("Datetime values must include an explicit timezone offset")
        return value

    @model_validator(mode="after")
    def validate_participants(self) -> BookingCreateRequest:
        if self.source == BookingSource.MEMBER_PORTAL and not self.participants:
            return self
        primary_participants = [
            participant for participant in self.participants if participant.is_primary
        ]
        if len(primary_participants) != 1:
            raise ValueError("exactly one primary participant is required")
        if primary_participants[0].participant_type == BookingParticipantType.GUEST:
            raise ValueError(
                "primary participant must be a member or staff participant in this phase"
            )
        if self.applies_to is not None:
            expected_applies_to = (
                BookingRuleAppliesTo.STAFF
                if primary_participants[0].participant_type == BookingParticipantType.STAFF
                else BookingRuleAppliesTo.MEMBER
            )
            if self.applies_to != expected_applies_to:
                raise ValueError(
                    "applies_to must match the primary participant bucket in this phase"
                )
        return self


class BookingCreateFailureDetail(BaseModel):
    code: str
    message: str
    field: str | None = None


class BookingCreateDecision(StrEnum):
    ALLOWED = "allowed"
    BLOCKED = "blocked"
    INDETERMINATE = "indeterminate"


class BookingParticipantSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    participant_type: BookingParticipantType
    person_id: uuid.UUID | None = None
    club_membership_id: uuid.UUID | None = None
    display_name: str
    guest_name: str | None = None
    sort_order: int
    is_primary: bool


class BookingSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID
    course_id: uuid.UUID
    tee_id: uuid.UUID | None = None
    start_lane: StartLane | None = None
    slot_datetime: datetime
    slot_interval_minutes: int
    status: BookingStatus
    source: BookingSource
    party_size: int
    primary_person_id: uuid.UUID | None = None
    primary_membership_id: uuid.UUID | None = None
    cart_flag: bool = False
    caddie_flag: bool = False
    fee_label: str | None = None
    payment_status: BookingPaymentStatus | None = None
    created_at: datetime
    updated_at: datetime
    participants: list[BookingParticipantSummary] = Field(default_factory=list)


class BookingCreateResult(BaseModel):
    decision: BookingCreateDecision
    booking: BookingSummary | None = None
    availability: AvailabilityPolicyResult | None = None
    failures: list[BookingCreateFailureDetail] = Field(default_factory=list)


class BookingLifecycleMutationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    booking_id: uuid.UUID
    acting_user_id: uuid.UUID


class BookingLifecycleMutationFailureDetail(BaseModel):
    code: str
    message: str
    field: str | None = None
    current_status: BookingStatus | None = None


class BookingLifecycleMutationDecision(StrEnum):
    ALLOWED = "allowed"
    BLOCKED = "blocked"


class BookingLifecycleMutationResult(BaseModel):
    booking_id: uuid.UUID
    decision: BookingLifecycleMutationDecision
    transition_applied: bool = False
    booking: BookingSummary | None = None
    failures: list[BookingLifecycleMutationFailureDetail] = Field(default_factory=list)


class BookingCancelRequest(BookingLifecycleMutationRequest):
    pass


class BookingCancelFailureDetail(BookingLifecycleMutationFailureDetail):
    pass


BookingCancelDecision = BookingLifecycleMutationDecision


class BookingCancelResult(BookingLifecycleMutationResult):
    booking_id: uuid.UUID
    decision: BookingCancelDecision
    failures: list[BookingCancelFailureDetail] = Field(default_factory=list)


class BookingCheckInRequest(BookingLifecycleMutationRequest):
    pass


class BookingCheckInFailureDetail(BookingLifecycleMutationFailureDetail):
    pass


BookingCheckInDecision = BookingLifecycleMutationDecision


class BookingCheckInResult(BookingLifecycleMutationResult):
    booking_id: uuid.UUID
    decision: BookingCheckInDecision
    failures: list[BookingCheckInFailureDetail] = Field(default_factory=list)


class BookingCompleteRequest(BookingLifecycleMutationRequest):
    pass


class BookingCompleteFailureDetail(BookingLifecycleMutationFailureDetail):
    pass


BookingCompleteDecision = BookingLifecycleMutationDecision


class BookingCompleteResult(BookingLifecycleMutationResult):
    booking_id: uuid.UUID
    decision: BookingCompleteDecision
    failures: list[BookingCompleteFailureDetail] = Field(default_factory=list)


class BookingNoShowRequest(BookingLifecycleMutationRequest):
    pass


class BookingNoShowFailureDetail(BookingLifecycleMutationFailureDetail):
    pass


BookingNoShowDecision = BookingLifecycleMutationDecision


class BookingNoShowResult(BookingLifecycleMutationResult):
    booking_id: uuid.UUID
    decision: BookingNoShowDecision
    failures: list[BookingNoShowFailureDetail] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Booking move
# ---------------------------------------------------------------------------


class BookingMoveInput(BaseModel):
    """HTTP request body for the move endpoint — booking_id comes from the URL path."""

    target_slot_datetime: datetime
    target_start_lane: StartLane | None = None
    target_tee_id: uuid.UUID | None = None

    @field_validator("target_slot_datetime")
    @classmethod
    def validate_timezone_aware(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("target_slot_datetime must include an explicit timezone offset")
        return value


class BookingMoveRequest(BaseModel):
    """Internal service request — assembled from URL path + body."""

    booking_id: uuid.UUID
    target_slot_datetime: datetime
    target_start_lane: StartLane | None = None
    target_tee_id: uuid.UUID | None = None

    @field_validator("target_slot_datetime")
    @classmethod
    def validate_timezone_aware(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("target_slot_datetime must include an explicit timezone offset")
        return value


class BookingMoveDecision(StrEnum):
    ALLOWED = "allowed"
    BLOCKED = "blocked"


class BookingMoveFailureDetail(BaseModel):
    code: str
    message: str
    field: str | None = None
    current_status: BookingStatus | None = None


class BookingMoveResult(BaseModel):
    booking_id: uuid.UUID
    decision: BookingMoveDecision
    transition_applied: bool = False
    booking: BookingSummary | None = None
    failures: list[BookingMoveFailureDetail] = Field(default_factory=list)

File: frontend/src/types/bookings.ts
export type BookingStatus = "reserved" | "checked_in" | "cancelled" | "completed" | "no_show";
export type BookingParticipantType = "member" | "guest" | "staff";
export type StartLane = "hole_1" | "hole_10";
export type BookingPaymentStatus = "pending" | "paid" | "complimentary" | "waived";

export interface BookingCreateParticipantInput {
  participant_type: BookingParticipantType;
  person_id?: string | null;
  guest_name?: string | null;
  is_primary: boolean;
}

export interface BookingCreateInput {
  course_id: string;
  tee_id?: string | null;
  start_lane?: StartLane | null;
  slot_datetime: string;
  slot_interval_minutes?: number | null;
  holes?: number | null;
  source?: "admin" | "member_portal";
  applies_to?: "member" | "guest" | "staff" | null;
  reference_datetime?: string | null;
  cart_flag?: boolean;
  caddie_flag?: boolean;
  participants: BookingCreateParticipantInput[];
}

export interface BookingParticipantSummary {
  id?: string;
  display_name: string;
  participant_type: BookingParticipantType;
  person_id?: string | null;
  club_membership_id?: string | null;
  guest_name?: string | null;
  sort_order?: number;
  is_primary: boolean;
}

export interface BookingSummary {
  id: string;
  club_id?: string;
  course_id?: string;
  tee_id?: string | null;
  start_lane?: StartLane | null;
  slot_interval_minutes?: number;
  holes: number;
  status: BookingStatus;
  source?: string;
  party_size: number;
  primary_person_id?: string | null;
  primary_membership_id?: string | null;
  cart_flag?: boolean;
  caddie_flag?: boolean;
  fee_label?: string | null;
  fee_amount?: string | null;
  fee_currency?: string | null;
  payment_status?: BookingPaymentStatus | null;
  slot_datetime: string;
  created_at?: string;
  updated_at?: string;
  participants: BookingParticipantSummary[];
}

export interface BookingCreateFailureDetail {
  code: string;
  message: string;
  field?: string | null;
}

export interface AvailabilityPolicyTrace {
  code: string;
  reason: string;
  details: Record<string, unknown>;
}

export interface AvailabilityPolicyNotice {
  code: string;
  message: string;
}

export interface AvailabilityPolicyResult {
  applies_to: "member" | "guest" | "staff";
  availability_status: string;
  blockers: AvailabilityPolicyTrace[];
  warnings: AvailabilityPolicyNotice[];
  resolved_checks: AvailabilityPolicyTrace[];
  unresolved_checks: AvailabilityPolicyTrace[];
}

export interface BookingCreateResult {
  decision: "allowed" | "blocked" | "indeterminate";
  booking: BookingSummary | null;
  availability: AvailabilityPolicyResult | null;
  failures: BookingCreateFailureDetail[];
}

export interface BookingUpdateInput {
  participants: BookingCreateParticipantInput[];
  holes?: number | null;
  applies_to?: "member" | "guest" | "staff" | null;
  reference_datetime?: string | null;
  cart_flag?: boolean;
  caddie_flag?: boolean;
}

export interface BookingUpdateFailureDetail {
  code: string;
  message: string;
  field?: string | null;
  current_status?: BookingStatus | null;
}

export interface BookingUpdateResult {
  booking_id: string;
  decision: "allowed" | "blocked" | "indeterminate";
  booking: BookingSummary | null;
  availability: AvailabilityPolicyResult | null;
  failures: BookingUpdateFailureDetail[];
}

export interface BookingFinanceMutationFailureDetail {
  code: string;
  message: string;
  field?: string | null;
  current_status?: BookingStatus | null;
  current_payment_status?: BookingPaymentStatus | null;
}

export interface BookingFinanceTransactionDetail {
  id: string;
  club_id: string;
  account_id: string;
  amount: string;
  type: "charge" | "payment" | "adjustment" | "refund";
  source: "booking" | "order" | "pos" | "manual";
  reference_id: string | null;
  description: string;
  created_at: string;
}

export interface BookingPaymentStatusUpdateInput {
  payment_status: BookingPaymentStatus;
}

export interface BookingPaymentStatusUpdateResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  update_applied: boolean;
  booking: BookingSummary | null;
  failures: BookingFinanceMutationFailureDetail[];
}

export interface BookingChargePostInput {
  amount?: string | null;
  description?: string | null;
}

export interface BookingChargePostResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  posting_applied: boolean;
  booking: BookingSummary | null;
  transaction: BookingFinanceTransactionDetail | null;
  balance: string | null;
  failures: BookingFinanceMutationFailureDetail[];
}

export interface BookingPaymentRecordResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  settlement_applied: boolean;
  booking: BookingSummary | null;
  transaction: BookingFinanceTransactionDetail | null;
  balance: string | null;
  failures: BookingFinanceMutationFailureDetail[];
}

export interface BookingRefundInput {
  amount?: string | null;
  description?: string | null;
}

export interface BookingRefundResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  refund_applied: boolean;
  booking: BookingSummary | null;
  transaction: BookingFinanceTransactionDetail | null;
  balance: string | null;
  failures: BookingFinanceMutationFailureDetail[];
}

export interface BookingLifecycleMutationFailureDetail {
  code: string;
  message: string;
  field?: string | null;
  current_status?: BookingStatus | null;
}

export interface BookingLifecycleMutationResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  transition_applied: boolean;
  booking: BookingSummary | null;
  failures: BookingLifecycleMutationFailureDetail[];
}

export interface BookingCancelFailureDetail extends BookingLifecycleMutationFailureDetail {}

export interface BookingCancelResult extends BookingLifecycleMutationResult {
  failures: BookingCancelFailureDetail[];
}

export interface BookingCheckInFailureDetail extends BookingLifecycleMutationFailureDetail {}

export interface BookingCheckInResult extends BookingLifecycleMutationResult {
  failures: BookingCheckInFailureDetail[];
}

export interface BookingCompleteFailureDetail extends BookingLifecycleMutationFailureDetail {}

export interface BookingCompleteResult extends BookingLifecycleMutationResult {
  failures: BookingCompleteFailureDetail[];
}

export interface BookingNoShowFailureDetail extends BookingLifecycleMutationFailureDetail {}

export interface BookingNoShowResult extends BookingLifecycleMutationResult {
  failures: BookingNoShowFailureDetail[];
}

export interface BookingMoveFailureDetail {
  code: string;
  message: string;
  field?: string | null;
  current_status?: BookingStatus | null;
}

export interface BookingMoveInput {
  target_slot_datetime: string;
  target_start_lane?: StartLane | null;
  target_tee_id?: string | null;
  participant_id?: string | null;
}

export interface BookingMoveResult {
  booking_id: string;
  decision: "allowed" | "blocked";
  transition_applied: boolean;
  booking: BookingSummary | null;
  failures: BookingMoveFailureDetail[];
}

export interface PlayerBookingReadModelItem {
  id: string;
  status: BookingStatus;
  source: "admin" | "member_portal";
  slot_datetime: string;
  holes: number;
  local_date: string;
  local_time: string;
  course_name: string;
  tee_name?: string | null;
  start_lane?: StartLane | null;
  party_size: number;
  primary_participant_name?: string | null;
  participant_names: string[];
  fee_label?: string | null;
  fee_amount?: string | null;
  fee_currency?: string | null;
  payment_status?: BookingPaymentStatus | null;
}

export interface PlayerBookingReadModelResponse {
  timezone: string;
  reference_datetime: string;
  upcoming: PlayerBookingReadModelItem[];
  history: PlayerBookingReadModelItem[];
}
